rust/src/operations/write.rs

```diff
@@ -526,6 +535,54 @@ mod tests {
         assert_eq!(table.get_file_uris().count(), 1)
     }
 
+    #[tokio::test]
+    async fn test_write_different_types() {
+        // Ensure write data is casted when data of a different type from the table is provided.
+        let schema = Arc::new(ArrowSchema::new(vec![Field::new(
+            "value",
+            arrow::datatypes::DataType::Int32,
+            true,
+        )]));
+
+        let batch = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![Arc::new(Int32Array::from(vec![Some(0), None]))],
+        )
+        .unwrap();
+        let table = DeltaOps::new_in_memory().write(vec![batch]).await.unwrap();
+
+        let schema = Arc::new(ArrowSchema::new(vec![Field::new(
+            "value",
+            arrow::datatypes::DataType::Utf8,
+            true,
+        )]));
+
+        let batch = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![Arc::new(StringArray::from(vec![
+                Some("Test123".to_owned()),
```

Comment:
I'm confused, where does this value go in the expected table? :confused:

Reply:
Right below, in line 570?

Reply:
Since it cannot be parsed as an int it will result in a null value. Which aligns with ansi sql but maybe we should only allow that if the user opts in?

Reply:
Oh nevermind. It seems like this is being inserted as null? That doesn't seem like what a user would want.

Reply:
Yeah I think we could disable that by passing `safe: true` in CastOptions: https://docs.rs/arrow-cast/40.0.0/arrow_cast/cast/struct.CastOptions.html